### PR TITLE
Helper methods to make full service usage easier

### PR DESCRIPTION
**Pull request by iamalwaysuncomfortable**
* Full service convenience methods
* Txo Pre-allocation logic to prevent errors in transaction sending

This pull request is WIP

### CODE REVIEW

File: fs_tester.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
+import logging
```
Comment (itdaniher) [inline fs_tester.py:1]: should this file go somewhere else?

File: forest/payments_monitor.py [hunk 2]
```diff
@@ -65,6 +65,8 @@
     get_invoice_by_amount="SELECT invoice_id, account FROM {self.table} WHERE unique_pmob=$1",
 )
 
+FEE = 4000000000
```
Comment (itdaniher) [inline forest/payments_monitor.py:68]: FEE_PMOB?

File: forest/payments_monitor.py [hunk 3]
```diff
@@ -235,66 +222,106 @@ async def get_balance(self) -> int:
...
-            await self.req(
-                {
-                    "method": "get_all_transaction_logs_for_account",
-                    "params": {"account_id": account_id},
-                }
-            )
-        )["result"]["transaction_log_map"]
-
-    async def build_single_txo_proposal(self, recipient: str, amount: str) -> dict:
+    async def get_signal_address(self, address: str = "") -> Optional[str]:
```
Comment (itdaniher) [inline forest/payments_monitor.py:225]: based, I have half-assed this every time I've need it. Glad to put that behind me.

File: forest/payments_monitor.py [hunk 4]
```diff
@@ -346,6 +371,191 @@ async def get_pending_transactions(self, from_block: int = 2) -> list[dict]:
...
+                    data.get("result", {})
+                    .get("wallet_status", {})
+                    .get("network_block_index", 0)
+                )
+            )
+        except:
+            start_block = 0
+        return start_block
+
+    async def get_all_txos_for_account(self, account_id: str = "") -> dict:
```
Comment (itdaniher) [inline forest/payments_monitor.py:405]: do we want to just return the unspent TXOs? if so, should this be get_all_utxos?

File: forest/payments_monitor.py [hunk 5]
```diff
@@ -346,6 +371,191 @@ async def get_pending_transactions(self, from_block: int = 2) -> list[dict]:
...
+            if status == "tx_status_failed":
+                logging.warning("tx failed, log: %s", tx_log)
+                return False
+            await asyncio.sleep(1)
+            time += 1
+        if time >= timeout:
+            logging.warning("Failed to confirm transaction in %s tries", timeout)
+        return False
+
+    async def cleanup_utxos(
```
Comment (itdaniher) [inline forest/payments_monitor.py:500]: hmmm, a single txo of specified size? not sure that's the right description

**Review by itdaniher — CHANGES_REQUESTED** [review]
looks pretty good overall, left some qs&cs!

File: forest/payments_monitor.py (hunk 5 above)
Comment (iamalwaysuncomfortable) [reply forest/payments_monitor.py:500]: This could be changed to a list of txos to cleanup to as well

File: forest/payments_monitor.py [hunk 6]
```diff
@@ -222,6 +331,85 @@ async def get_balance(self) -> int:
...
+            value_pairs = [(v[0], str(v[1])) for v in addresses_and_values]
+            params["addresses_and_values"] = value_pairs
+        if input_txo_ids:
+            params["input_txo_ids"] = input_txo_ids
+        if submit and comment:
+            params["comment"] = comment
+        if not submit and log:
+            params["log_tx_proposal"] = log
+        if comment:
+            params["comment"] = comment
```
Comment (technillogue) [inline forest/payments_monitor.py:408]: is this right?

**Review by iamalwaysuncomfortable — COMMENTED** [review]
Changes done

**Review by technillogue — APPROVED** [review]
I haven't fully checked that this won't break anything currently used, but I think this looks good

File: forest/payments_monitor.py [hunk 7]
```diff
@@ -252,7 +235,86 @@ async def get_balance(self) -> int:
...
+        input_txo_ids: Optional[list[str]] = None,
+        submit: bool = False,
+        log: bool = False,
+        comment: str = "",
+    ) -> dict:
+        """
+        Build a single_txo or multi_txo transaction. If building a single txo
+        transaction use value_pmob and recipient_public_address args. If
+        building a multi-txo transaction, use the addresses_and_values arg to
+        specify a list of [mobilecoin address, amount] pairs.
```
Comment (technillogue) [inline forest/payments_monitor.py:274]: should this be two methods?

**Pull request closed without merging**